crates/interpreter/src/module.rs

```diff
@@ -222,27 +214,45 @@ impl<'m> Module<'m> {
         unreachable!()
     }
 
-    pub(crate) fn skip(
-        &mut self, parser: &mut Parser<'m>, depth: LabelIdx,
-        compute: impl Fn(&mut Parser<'m>, LabelIdx),
-    ) {
-        let saved = parser.save();
-        let key = CacheKey::Skip { ptr: saved.as_ptr(), depth };
-        match self.cache.get(&key) {
-            Some(delta) => unsafe { parser.restore(&saved[delta.skip ..]) },
-            None => {
-                compute(parser, depth);
-                let delta = saved.len() - parser.save().len();
-                self.cache.put(key, CacheValue { skip: delta });
+    pub(crate) fn jump_from_if(&mut self, parser: &mut Parser<'m>) {
+        let (i, j) = self.side_table_indices;
+        let entry = self.side_tables[i][j].view();
+        let delta = entry.delta_ip as isize - 1;
+        unsafe {
+            parser.restore(Self::jump(parser.save(), delta));
+            let saved = parser.save();
+            if parser.parse_instr().unwrap() == Instr::End {
```

Comment:
Could you explain more? An example would be appreciated.

I basically followed the treatment here: https://github.com/google/wasefire/blob/main/crates/interpreter/src/parser.rs#L566-L569

Reply:
Your observation looks very relevant, and I'd really like to understand it better.

The error comes from [this function](https://github.com/zhouwfang/wasefire/blob/54b17f7c291cd48b576e74defb668bab7abf12f6/target/wasefire/applet.wast#L16101), and its side table is
```
2024-12-02T21:57:57.5343889Z i=53
2024-12-02T21:57:57.5343957Z j=0
2024-12-02T21:57:57.5344023Z delta_ip=14
2024-12-02T21:57:57.5344094Z delta_stp=1
2024-12-02T21:57:57.5344155Z val_cnt=0
2024-12-02T21:57:57.5344219Z pop_cnt=0
2024-12-02T21:57:57.5344285Z j=1
2024-12-02T21:57:57.5344353Z delta_ip=8
2024-12-02T21:57:57.5344422Z delta_stp=1
2024-12-02T21:57:57.5344484Z val_cnt=0
2024-12-02T21:57:57.5344547Z pop_cnt=0
```

The error is `panicked at /home/runner/work/wasefire/wasefire/crates/interpreter/src/module.rs:243:40:
index out of bounds: the len is 2 but the index is 2`

Reply:
The side table looks consistent, there's indeed 13 bytes in the first if and 7 in the second one. You could print the `parser.save()` before each instruction (including the instruction debug) and see how the jump is taken. We should ideally land right after the `End`.

Note that if you don't jump at the correct place, you may interpret arbitrary code (including code that was not present, due to WASM not having aligned bytecode). For example, if you jump on the byte preceding the `End` of the first if, you'll see another if because of the `4` of `local.set 4`, and for the second if you'll see `br` because of the 12 (offset of `i32.store`).